_scala_items/7-education.md

```diff
@@ -0,0 +1,83 @@
+---
+shortTitle: "Ideal for teaching"
+shortDescription: "Scala is ideal for teaching programming to beginners as well as for teaching advanced software engineering courses."
+---
+
+<div class="wrap">
+  <div class="scala-text scala-text-large">
+    <h3>Concise, Readable, and Polyvalent</h3>
+    <p>
+      Most of the concepts involved in software design directly map
+      into Scala constructs. The concise syntax of Scala allows the teachers
+      and the learners to focus on those interesting concepts without dealing
+      with tedious low-level implementation issues.
+    </p>
+    <p>
+      The example in file <code>HelloWorld.scala</code> below shows how a “hello
+      world” program looks like in Scala. In <code>Modeling.scala</code>, we show an
+      example of structuring the information of a problem domain in Scala. In
+      <code>Modules.scala</code>, we show how to use classes to implement
+      software modules. Last, in <code>Algorithms.scala</code>, we show how the
+      standard Scala collections can be leveraged to implement algorithms with
+      few lines of code.
+    </p>
+    <p>
+      Learn more in the dedicated page about
+      <a href="{% link teach.md %}">Teaching</a>.
+    </p>
+  </div>
+
+  <div class="scala-code">
+
+    <div class="code-element">
+      <div class="bar-code"><span>HelloWorld.scala</span></div>
+      <pre><code>@main def run() = println("Hello, World!")</code></pre>
+    </div>
+
+    <div class="code-element">
+      <div class="bar-code"><span>Modules.scala</span></div>
```

Comment:
I initially wrote something that was using abstract methods and abstract type members, but I thought maybe a simpler example would also work…
Indeed, working with classes as shown in that example is not outstanding, but my idea was just to show that it’s simple. The same in Java would be much more complicated…

~~~ java
class DatabaseAccess {

  private final Connection connection;

  public DatabaseAccess(Connection connection) {
    this.connection = connection;
  }

  public List<Data> readData() { ... }

}

class HttpServer {

  private final DatabaseAccess databaseAccess;

  public HttpServer(DatabaseAccess databaseAccess) {
    this.databaseAccess = databaseAccess;
  }

}
~~~

(or maybe we can do it in a simpler way in Java?

Reply:
It's not more complicated in Java. It's only more verbose, because transferring constructor parameters to fields requires some code. But that has nothing to do with "modules". It's just the regular verbosity of a Java class.

Reply:
Okay, so maybe “more complicated” was not the right wording, what about “less readable”?

Do you think we should show something more spectacular? Or is it okay to go with a straightforward example?

Reply:
If you want to talk about *modules*, then I think we would need something more "spectacular". But it doesn't need to be about modules. It can be an example about "no boilerplate" (perhaps also showing a `val` field).